staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered.go

```diff
@@ -0,0 +1,259 @@
+				// Note, that in case of a panic, the return value will be false and
+				// the loop execution will continue.
+				defer runtime.HandleCrash()
+
+				events := b.collectLastEvents()
+				b.processEvents(events)
+				return len(events) == 0
+			}(); last {
+				break
+			}
+		}
+	}()
+	return nil
+}
+
+func (b *bufferedBackend) Shutdown() {
+	<-b.shutdownCh
+
+	// Write locking reqMutex will guarantee that all requests will be completed
+	// by the time the goroutine continues the execution. Since this line is
+	// executed after shutdownCh was closed, no new requests will follow this
+	// lock, because read lock is called in the same goroutine that closes
+	// shutdownCh before exiting.
+	b.reqMutex.Lock()
+	b.reqMutex.Unlock()
+}
+
+// runWorkerRoutine runs a loop that collects events from the buffer. When
+// stopCh is closed, runWorkerRoutine stops and closes the buffer.
+func (b *bufferedBackend) runWorkerRoutine(stopCh <-chan struct{}) {
+	defer close(b.buffer)
+
+	for {
+		func() {
+			// Recover from any panics caused by this function so a panic in the
+			// goroutine can't bring down the main routine.
+			defer runtime.HandleCrash()
+
+			t := time.NewTimer(b.maxBatchWait)
+			defer t.Stop() // Release ticker resources
+
+			b.processEvents(b.collectEvents(stopCh, t.C))
+		}()
+
+		select {
+		case <-stopCh:
+			return
+		default:
+		}
+	}
+}
+
+// collectEvents attempts to collect some number of events in a batch.
+//
+// The following things can cause collectEvents to stop and return the list
+// of events:
+//
+//   * Some maximum number of events are received.
+//   * Timer has passed, all queued events are sent.
+//   * StopCh is closed, all queued events are sent.
+//
+func (b *bufferedBackend) collectEvents(stopCh <-chan struct{}, timer <-chan time.Time) []*auditinternal.Event {
+	var events []*auditinternal.Event
+
+L:
+	for i := 0; i < b.maxBatchSize; i++ {
+		select {
+		case ev, ok := <-b.buffer:
+			// Buffer channel was closed and no new events will follow.
+			if !ok {
+				break L
+			}
+			events = append(events, ev)
+		case <-timer:
+			// Timer has expired. Send whatever events are in the queue.
+			break L
+		case <-stopCh:
+			// Webhook has shut down. Send the last events.
+			break L
+		}
+	}
+
+	return events
+}
+
+// collectLastEvents assumes that the buffer was closed. It collects the first
+// maxBatchSize events from the closed buffer into a batch and returns them.
+func (b *bufferedBackend) collectLastEvents() []*auditinternal.Event {
+	var events []*auditinternal.Event
+
+	for i := 0; i < b.maxBatchSize; i++ {
+		ev, ok := <-b.buffer
+		if !ok {
+			break
+		}
+		events = append(events, ev)
+	}
+
+	return events
+}
+
+// Func processEvents process the batch events in a goroutine use the
+// real backend interface's ProcessEvents.
+func (b *bufferedBackend) processEvents(events []*auditinternal.Event) {
+	if len(events) == 0 {
+		return
+	}
+
+	if b.throttle != nil {
+		b.throttle.Accept()
+	}
```

Comment:
I think we should have a limit on the number of active goroutines in addition to the throttle. For example, if the backend is locked up, we don't want to spawn infinite goroutines.

Reply:
do not seem to occur `if the backend is locked up, we don't want to spawn infinite goroutines.`
Because the lock is rwLock, and only Lock in shutdown can block RLock, but Lock can only occur after Run exit.

Reply:
`b.throttle.Accept()` in `processEvents()` happens before `b.reqMutex.Lock()` in `Shutdown()`, it is guaranteed by `b.shutdownCh`. So `spawn infinite goroutines` by no means to occur.

Reply:
I'm confused by your response. What does this have to do with shutdown? This scenario could occur without shutting down.

Reply:
ping.